Import the pandas, NumPy and os libraries.

In [2]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Create data path
path = r"C:\Users\Jason\Instacart Basket Analysis"

Importing 'orders.csv', 'products.csv' and 'departments_data.csv' as df_ords, df_prods and df_dep.

In [4]:
# orders
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
# products
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
# departments
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'departments.csv'), index_col = False)

4.4 Wrangling Procedures

In [5]:
# Dropping eval_set column from orders.csv
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [6]:
# Changing the variable data type
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [7]:
# Overwriting df with new transposed df
df_dep_t = df_dep.T

In [8]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [9]:
# Create a new header
new_header = df_dep_t.iloc[0]

In [10]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [11]:
# Create the new header
df_dep_t_new.columns = new_header

In [12]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [13]:
# Creating a data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [14]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

Beginning of Task 4.4 steps

In [15]:
df_ords.head()

order_id user_id eval_set  order_number  order_dow  order_hour_of_day  \
0  2539329       1    prior             1          2                  8   
1  2398795       1    prior             2          3                  7   
2   473747       1    prior             3          3                 12   
3  2254736       1    prior             4          4                  7   
4   431534       1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [16]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(3)
memory usage: 182.7+ MB


2. Changing order_number to an object, not needed for mathematical analysis.

In [17]:
df_ords['order_number'] = df_ords['order_number'].astype(str)

In [18]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   eval_set                object 
 3   order_number            object 
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(2), object(4)
memory usage: 182.7+ MB


3. Changing 'order_dow' to 'orders_day_of_week'

In [19]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [20]:
df_ords.head()

order_id user_id eval_set order_number  orders_day_of_week  \
0  2539329       1    prior            1                   2   
1  2398795       1    prior            2                   3   
2   473747       1    prior            3                   3   
3  2254736       1    prior            4                   4   
4   431534       1    prior            5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

4. Find the busiest hour for placing orders. 10am is the busiest hour for the placing of orders.

In [21]:
df_ords['order_hour_of_day'].value_counts(dropna = True)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

5. Value of 4 in the 'department_id' = 'produce'

In [22]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

6. Breakfast item sales

In [23]:
# Creating a subset for breakfast item sales
df_prods_breakfast =  df_prods[df_prods['department_id']==14]

In [24]:
df_prods[df_prods['department_id']==14]

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [25]:
df_prods_breakfast =  df_prods[df_prods['department_id']==14]

In [26]:
df_prods_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

7. Dinner party product details - alcohol, deli, beverages and meat/seafood

In [27]:
# Creating a subset for dinner party
df_prods_dinner_party =  df_prods[df_prods['department_id'].isin([5,7,12,20])]

In [28]:
df_prods[df_prods['department_id'].isin([5,7,12,20])]

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [29]:
df_prods_dinner_party.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

8. How many rows does the last dataframe you created have? - 7650 rows, 5 columns.

In [30]:
df_prods_dinner_party.shape

(7650, 5)

9. All information about 'user_id' of '1'

In [31]:
# Creating a dataframe for user_id_1
df_user_id_1=df_ords.loc[df_ords['user_id'] == '1']

In [32]:
df_user_id_1

order_id user_id eval_set order_number  orders_day_of_week  \
0   2539329       1    prior            1                   2   
1   2398795       1    prior            2                   3   
2    473747       1    prior            3                   3   
3   2254736       1    prior            4                   4   
4    431534       1    prior            5                   4   
5   3367565       1    prior            6                   2   
6    550135       1    prior            7                   1   
7   3108588       1    prior            8                   1   
8   2295261       1    prior            9                   1   
9   2550362       1    prior           10                   4   
10  1187899       1    train           11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

10. user_id_1 details on user's behaviour. = User prefers to order early in the week, 2.64 indicates Tuesday or Wednesday. Orders early in the day, median is 8am for orders, average is 10am.

In [33]:
df_user_id_1.describe()

orders_day_of_week  order_hour_of_day  days_since_prior_order
count           11.000000          11.000000               10.000000
mean             2.636364          10.090909               19.000000
std              1.286291           3.477198                9.030811
min              1.000000           7.000000                0.000000
25%              1.500000           7.500000               14.250000
50%              3.000000           8.000000               19.500000
75%              4.000000          13.000000               26.250000
max              4.000000          16.000000               30.000000

12. Export your df_ords dataframe as “orders_wrangled.csv”

In [34]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

13. Export the df_dep_t_new dataframe as “departments_wrangled.csv”

In [36]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))